Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/OptimisticDirectExecution/OptimisticDirectExecutionQueryPipelineStage.cs

```
@@ -8,172 +8,306 @@ namespace Microsoft.Azure.Cosmos.Query.Core.Pipeline.OptimisticDirectExecutionQu
-            QueryPartitionRangePageAsyncEnumerator queryPartitionRangePageAsyncEnumerator)
-        {
-            this.queryPartitionRangePageAsyncEnumerator = queryPartitionRangePageAsyncEnumerator ?? throw new ArgumentNullException(nameof(queryPartitionRangePageAsyncEnumerator));
-        }
-
-        public TryCatch<QueryPage> Current { get; private set; } 
-
-        public ValueTask DisposeAsync()
+        //private readonly Func<IQueryPipelineStage> queryPipelineStageFactory;
+        private TryCatch<IQueryPipelineStage> inner;
+        //private TryCatch<IQueryPipelineStage> currentPipeline;
+        private QueryPartitionRangePageAsyncEnumerator partitionPageEnumerator;
+        private static DocumentContainer mDocumentContainer;
+        private static CosmosQueryContext mCosmosQueryContext;
+        private static InputParameters mInputParameters;
+        private static FeedRangeEpk mTargetRange;
+        
+        public OptimisticDirectExecutionQueryPipelineStage(TryCatch<IQueryPipelineStage> monadicCreate)
         {
-            return this.queryPartitionRangePageAsyncEnumerator.DisposeAsync();
+            this.inner = monadicCreate;
         }
 
-        public void SetCancellationToken(CancellationToken cancellationToken)
-        {
-            this.queryPartitionRangePageAsyncEnumerator.SetCancellationToken(cancellationToken);
-        }
+        public TryCatch<QueryPage> Current { get; private set; }
 
+        public ValueTask DisposeAsync() => default;
+       
         public async ValueTask<bool> MoveNextAsync(ITrace trace)
         {
-            if (trace == null)
+            bool result = await this.inner.Result.MoveNextAsync(trace);
+            bool isGoneException = this.inner.Result.Current.Failed
+                && this.inner.Result.Current.InnerMostException is CosmosException exception
+                && exception.StatusCode == System.Net.HttpStatusCode.Gone
+                && exception.SubStatusCode == 1002;
```

This seems flaky.

QueryPage should own the decision about whether something failed with GoneException instead of every pipeline copying this logic independently.

---

Just checking for GoneException does not have much value (as it is currently a one line code anyways) and if we add sub status code to the method then we would also have to create a similar method for all different sub status codes related to GoneException (1007 : completingSplit)

---

Furthermore, no other status code checking functions are exposed by QueryPage, so why should only GoneException get one